Printf - Ordering in Buffers using 1 atomic inst

aaronenyeshi opened the pull request:
Previously, the packets were being allocated into the two buffers in different orders with respect to each other. Now using a single atomic compare_exchange_weak, we can maintain ordering in both buffers with respect with each other. Also, string arguments in a single printf, will be allocated consecutively in-order as well. Offsets will be managed in a single ullong, which can be divided into 2 4byte uints. Also, we are now performing a look-ahead to check for the space required to allocation into the string buffer, so we can perform space allocation in printf command chunks.

Also small update to test.

aaronenyeshi committed:
Printf - Ordering in Buffers using 1 atomic inst

Previously, the packets were being allocated into the two buffers in different orders with respect to each other. Now using a single atomic compare_exchange_weak, we can maintain ordering in both buffers with respect with each other. Also, string arguments in a single printf, will be allocated consecutively in-order as well. Offsets will be managed in a single ullong, which can be divided into 2 4byte uints. Also, we are now performing a look-ahead to check for the space required to allocation into the string buffer, so we can perform space allocation in printf command chunks.

scchan commented on include/hc_printf.hpp:
```diff
@@ -111,56 +102,93 @@ static inline void copy_n(char* dest, const char* src, unsigned int len) [[hc,cp
+	  old_long = old_off.al.load();
+	  try_off.ai[0] = old_off.ai[0] + count_arg + 1;
+	  try_off.ai[1] = old_off.ai[1] + count_char;
+    } while(!(queue[3].data.al.compare_exchange_weak(old_long, (unsigned long long)try_off.al.load())));
```
old_off and try_off are local to the thread and therefore they don't need to be defined with an atomic type

scchan commented on include/hc_printf.hpp:
```diff
@@ -64,40 +64,31 @@ enum PrintfError {
+    // Combine offsets into 1 atomic element to maintain order and atomicity
+    printfBuffer[3].type = PRINTF_OFFSETS;
+    printfBuffer[3].data.ai[0] = 4;
+    printfBuffer[3].data.ai[1] = 0;
```
Could we create names for different header indices?

aaronenyeshi replied:
Sorry, what did you mean create names? Right now, different header indices have a type member that aligns with enum PrintfPacketDataType.

scchan replied:
right... could we re-arrange the PrintPacketDataType enum to make the code more readable:

```
enum PrintfPacketDataType {
  PRINTF_BUFFER_SIZE = 0,
  PRINTF_STRING_BUFFER,
...
};

...

printfBuffer[PRINTF_BUFFER_SIZE].type = PRINTF_BUFFER_SIZE;
printfBuffer[PRINTF_BUFFER_SIZE].data.ui = numElements;
```

scchan commented on include/hc_printf.hpp:
```diff
@@ -111,56 +102,93 @@ static inline void copy_n(char* dest, const char* src, unsigned int len) [[hc,cp
+    if (poffset + count_arg + 1 > queue[0].data.ui) {
+      error = PRINTF_BUFFER_OVERFLOW;
+    }
+	else if (soffset + count_char > queue[2].data.ui){
```
indent issue...

aaronenyeshi committed:
Printf - Small fixes to offset implementation

Won't use atomic type for local ullong and uints. Fix indentation issues. Re-arrange PrintfPacketDataType enum, and added comments to make code more readable.

aaronenyeshi commented:
I've fixed the implementation due to comments. Runs correctly on the testcases. @whchung can you take a look while @scchan is away

whchung commented:
@aaronenyeshi will do.

whchung commented on include/hc_printf.hpp:
```diff
@@ -64,40 +71,33 @@ enum PrintfError {
+    printfBuffer[1].type = PRINTF_STRING_BUFFER;
+    printfBuffer[1].data.ptr = hc::am_alloc(sizeof(char) * numElements * 12, a, amHostCoherent);
+    printfBuffer[2].type = PRINTF_STRING_BUFFER_SIZE;
+    printfBuffer[2].data.ui = numElements * 12;
```
@aaronenyeshi  i'm a bit afraid of all these magic numbers. could you try change them into enums?

whchung commented on include/hc_printf.hpp:
```diff
@@ -241,14 +276,14 @@ static inline void processPrintfBuffer(PrintfPacket* gpuBuffer) {
-  unsigned int cursor = gpuBuffer[1].data.ui;
-  unsigned int numPackets = ((bufferSize<cursor)?bufferSize:cursor) - 5;
+  unsigned int cursor = gpuBuffer[3].data.ai[0];
+  unsigned int numPackets = ((bufferSize<cursor)?bufferSize:cursor) - 4;
```
@aaronenyeshi  i'm a bit afraid of all these magic numbers. could you try change them into enums?

whchung reviewed (changes requested):
Could you change those magic values into enums?

aaronenyeshi committed:
Printf - use enums for header values

Switch to using enums with values for header indicies instead of magic numbers.

aaronenyeshi commented:
@whchung I've removed the magic values and used enums instead. I didn't know this was what SiuChi meant initially. Thanks for reviewing!

whchung commented on include/hc_printf.hpp:
```diff
@@ -23,23 +23,31 @@ union PrintfPacketData {
-  std::atomic_int ai;
+
+  // Used by headers for atomic offset insts
+  std::atomic_uint ai[2];
```
@aaronenyeshi for other types they are relatively easy to understand, but perhaps it's better to document what these fields are for. especially what `ai[0]` `ai[1]` `uia[0]` `uia[1]` means.

whchung reviewed (changes requested):
@aaronenyeshi looks much better now, if you could add some more comments on `PrintfPacketData`.

aaronenyeshi committed:
Printf - simplify atomic offset PrintfPacketData type

Removed ai[2] since we can access offsets individually using unsigned int uia[2]. Using two 4byte uint offsets in uia[2], and perform updates to these values in-order using single atomic 8byte ullong al. Refactor code to use uia instead of ai.

aaronenyeshi commented:
@whchung I've changed the atomic data members a little bit. Seems that we didn't need atomic uint anymore, since we only use a single atomic compare_exchange on atomic ullong al. I've added some comments as well.

whchung commented on include/hc_printf.hpp:
```diff
@@ -64,40 +76,34 @@ enum PrintfError {
 static inline PrintfPacket* createPrintfBuffer(hc::accelerator& a, const unsigned int numElements) {
   PrintfPacket* printfBuffer = NULL;
-  if (numElements > 6) {
+  if (numElements > 5) {
```
5 looks to be a magic value worthwhile be enumed.

also the logic would return a `nullptr`, but other APIs such as `deletePrintBuffer` doesn't guard `nullptr`. I would recommend you improve the test case to check `nullptr` cases as well.

aaronenyeshi replied:
Should test case also check for error cases from printf call on gpu?

whchung replied:
since the test case use FileCheck to check its output perhaps it's fine just to add nullptr check to make sure the tests won't crash, but could capture the case when the implementation returns a nullptr.

aaronenyeshi replied:
Do you mean use an assertion for the case where create returns a nullptr? Or another sort of capture?

whchung replied:
since you use FileCheck to check the output i think it's not good to let the test case crash. so you can catch nullptr and output a different message, and use `CHECK-NOT`

aaronenyeshi replied:
I've committed a change for the test cases to check for nullptr. Just a note, printf, processprintf, and deleteprintf should be guarding against nullptr buffer. So, test case will fail by `CHECK-DAG`, since the test won't crash if createprintf fails.

whchung commented on include/hc_printf.hpp:
```diff
@@ -23,23 +23,35 @@ union PrintfPacketData {
+  // the Printf string buffer offset uia[1]. We want to update a
+  // single atomic offset al, so that both buffers will increment
+  // in order with respect to each other. This way, offsets can be
+  // managed by single ullong 8B, which is divided into 2 4B uints.
```
@aaronenyeshi description here doesn't seem to clearly explain the relationship between `al`, `uia[]`, and `ull`. there is no explaination about what `ull` would be used.

whchung commented on include/hc_printf.hpp:
```diff
@@ -111,56 +117,91 @@ static inline void copy_n(char* dest, const char* src, unsigned int len) [[hc,cp
+      old_off.ull = queue[PRINTF_OFFSETS].data.al.load();
+      try_off.uia[0] = old_off.uia[0] + count_arg + 1;
+      try_off.uia[1] = old_off.uia[1] + count_char;
+    } while(!(queue[PRINTF_OFFSETS].data.al.compare_exchange_weak(old_off.ull, try_off.ull)));
```
@aaronenyeshi this relates to my comments earlier in the file. I think the relationship between `ull`, `uia[]`, `al` deserved to be explained a bit better.

aaronenyeshi committed:
Printf - add guards for functions and tests

Ensure that all function calls to printf functions will have guards to check for nullptr. Also test cases will check if creation was successful. Changes to comments about atomic variable usage added.
